### DISCUSSION TITLE
Portable版とINI形式について

### DISCUSSION THREAD

**Discussion by sayurin** [ffftp/ffftp#249] [General]
Portable版はその名の通りポータブルな利用を目的としている。つまり、USBメモリに入れて別マシンに移したりできる。
それを実現するために、設定情報をレジストリでなくINI形式として保存することをサポートしている。
しかし、レジストリを使いたくないという自己満足のために、INI形式を選択する人が一定数いるように感じられる。
その結果 #30 のような問題も発生している。

どういった機能が求められているのかはっきりさせたい。

**Comment by itagagaki** [comment]
#30 がユーザーの自己満足のせいとおっしゃっているように見えますが、それはちょっと違和感があります。
一般ユーザーがINI形式を選択する理由は何だってよくないですか？

そして、一般ユーザーにとって、
・INI形式とレジストリのどちらを選択するか
・Portable版と.msi版のどちらでインストールするか
は別のトピックです。

.msi版でINI形式が使えないのは、ユーザーのせいではないと思います。

開発者の都合で、インストール方法によって使える機能が違ってしまうのなら、
「INI形式を選択したければポータブル版をダウンロードしてインストールしてください」
という注意書きが、せめて必要なのではないかと思います。

**Reply by sayurin** [reply]
少し立場が異なります。INI形式はポータブル版を実現するための手段と考えています。例えばレジストリ形式であっても設定のバックアップ機能は提供しています。そのため「一般ユーザーがINI形式を選択する」は手段の目的化と捉えています。

---

辿れる範囲で歴史を遡ると

https://github.com/ffftp/ffftp/blob/b15217f71738c0a90261e0e931f7afc98089f288/doc/history.ja.txt#L699-L705

ということで、当初INI形式だったものの、マルチユーザー対応のためにレジストリ形式が導入されたようです。
なので、Portable版以外の目的でINIを選択するというのは、懐古なのかなと。

---

動作環境とINI書き込み可否の関係を整理するとこんな感じ。

- インストール先を書き込み可能なフォルダに変更した場合
  → INIを書き込み**可能**
- 既定のProgram Filesにインストールした場合
  - Windows 9xの場合
    → INIを書き込み**可能**（アクセス制御がないため）
  - Windows xp以前の場合
    - 標準ユーザーの場合
      → INIを書き込み**不可**（権限がないため）
    - Administrators権限の場合
      → INIを書き込み**可能**
  - Windows Vista以降の場合
    - ffftp 32bit版の場合
      → INIを書き込み**可能**（Windows 9x互換のファイルシステム仮想化のため）
    - ffftp 64bit版の場合
      → INIを書き込み**不可**（権限がないため）

---

新たな選択肢として、一般的なインストーラーによくある「このコンピュータを使用するすべてのユーザー」「自分のみ」というインストールタイプを提供するのもひとつの方法かもしれません。（現状は、強制的に前者）
「自分のみ」とした場合、自分が書き込み可能なディレクトリ （既定では`%LOCALAPPDATA%`＝`C:\Users\<UserName>\AppData\Local`）にインストールされるので、INIも書き込み可能になります。

これは [junerさんの提案](https://github.com/ffftp/ffftp/issues/30#issuecomment-385580714)でもあります。

**Comment by itagagaki** [comment]
それの議論はあくまでも開発者視点であって、一般ユーザーのことを考えることが抜け落ちていると感じます。
一般ユーザーが、たとえば窓の杜でFFFTPをダウンロードすると、選択の余地なく.msi版がダウンロードされます。
それでインストールし、目的は何であれINI形式を選択すると、一般ユーザーにはわけのわからないエラーとなるわけです。
ですから、せめて
・64bit .msiだとINIは使えませんよという情報
・どうしてもINIを使いたいならこういう手がありますよ（たとえば***からPortable版をダウンロードしてインストール）という情報
・なぜINIよりもレジストリが推奨されるか
といった情報は必要だとは思いませんか？

**Reply by sayurin** [reply]
そのような説明が求められる機能だったとすると、MSI版からINI形式対応の機能そのものを削った方がいいですねぇ…。

説明したところで、窓の杜のような二次配布サイトで情報ロストされるようなら制御できませんし（現にポータブル版が配布されていないとのことなら）。

**Reply by itagagaki** [reply]
たしかに、ごもっともです。

エラーダイアログが出たときにユーザーが困ってしまうのは別の問題なので、[新しいイシュー](https://github.com/ffftp/ffftp/issues/250#issue-817075187)を追加しました。
